src/java/org/apache/cassandra/cql3/InMarker.java

```diff
@@ -0,0 +1,178 @@
+import org.apache.cassandra.db.marshal.ByteBufferAccessor;
+import org.apache.cassandra.db.marshal.CollectionType;
+import org.apache.cassandra.db.marshal.ListType;
+import org.apache.cassandra.db.marshal.MapType;
+import org.apache.cassandra.db.marshal.SetType;
+import org.apache.cassandra.db.marshal.UserType;
+import org.apache.cassandra.db.marshal.TupleType;
+import org.apache.cassandra.exceptions.InvalidRequestException;
+import org.apache.cassandra.utils.ByteBufferUtil;
+
+public class InMarker extends Terms.NonTerminals
+{
+    private final int bindIndex;
+    private final ColumnSpecification receiver;
+
+    protected InMarker(int bindIndex, ColumnSpecification receiver)
+    {
+        this.bindIndex = bindIndex;
+        this.receiver = receiver;
+    }
+
+    @Override
+    public void addFunctionsTo(List<Function> functions) {}
+
+    @Override
+    public void collectMarkerSpecification(VariableSpecifications boundNames)
+    {
+        boundNames.add(bindIndex, receiver);
+    }
+
+    @Override
+    public Terminals bind(QueryOptions options)
+    {
+        ByteBuffer values = options.getValues().get(bindIndex);
+        if (values == null)
+            return null;
+
+        if (values == ByteBufferUtil.UNSET_BYTE_BUFFER)
+            return UNSET_TERMINALS;
+
+        ListType<?> type = (ListType) receiver.type;
+        return toTerminals(values, type, terminalConverter(type.getElementsType()));
+    }
+
+    private <T> Terminals toTerminals(ByteBuffer value,
+                                      ListType<T> type,
+                                      java.util.function.Function<ByteBuffer, Term.Terminal> terminalConverter)
+    {
+        List<T> l = type.getSerializer().deserialize(value, ByteBufferAccessor.instance);
+        List<Term.Terminal> terminals = new ArrayList<>(l.size());
+        for (T element : l)
+        {
+            terminals.add(element == null ? null : terminalConverter.apply(type.getElementsType().decompose(element)));
+        }
+        return Terminals.of(terminals);
+    }
+
+    public static java.util.function.Function<ByteBuffer, Term.Terminal> terminalConverter(AbstractType<?> type)
+    {
+        if (type.isCollection())
+        {
+            switch (((CollectionType<?>) type).kind)
+            {
+                case LIST:
+                    return e -> Lists.Value.fromSerialized(e, (ListType<?>) type);
+                case SET:
+                    return e -> Sets.Value.fromSerialized(e, (SetType<?>) type);
+                case MAP:
+                    return e -> Maps.Value.fromSerialized(e, (MapType<?, ?>) type);
+            }
+            throw new AssertionError();
+        }
+
+        if (type.isUDT())
+            return e -> UserTypes.Value.fromSerialized(e, (UserType) type);
+
+        if (type.isTuple())
+            return e -> Tuples.Value.fromSerialized(e, (TupleType) type);
+
+        return Constants.Value::new;
```

Comment:
Does vector type return this too ?

Reply:
The vector type is now a `MultiElements` type